File: event.server.csharp/ConnectionManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using EventManager.Service;
using ModelComponent.Domain;
using ModelComponent.Model;
using Thrift.Protocol;
using Thrift.Transport;

namespace @event.server.csharp
{
    public class ConnectionManager: Subsciption.Iface
    {
        private readonly List<Subscriber.Client> _subscribers;

        private readonly ModelPlayer _modelPlayer;
        private readonly ModelEvent _modelEvent;
        private readonly ModelUser _modelUser;

        public ConnectionManager()
        {
            _subscribers = new List<Subscriber.Client>();
            _modelEvent = new ModelEvent();   
            _modelPlayer = new ModelPlayer();
            _modelUser = new ModelUser();
        }

        private UserData ConvertToUserData(User user)
        {
            return new UserData(user.Name, user.Password);   
        }

        private User ConvertToUser(UserData user)
        {
            return new User(user.Id, user.Username, user.Password);
        }

        private PlayerData ConvertToPlayerData(Player player)
        {
            return new PlayerData(player.Name, player.Age);
        }

        private Player ConvertToPlayer(PlayerData player)
        {
            return new Player(player.Id, player.Name, player.Age);
        }

        private EventData ConvertToEventData(Event element)
        {
            return new EventData(element.Distance, element.Style.ToString());
        }

        private Event ConvertToEvent(EventData element)
        {
            return new Event(element.Id, element.Distance, EventStyle.FromString(element.Style));
        }

        public string sendMessage(string message)
        {
            Console.Write(message);
            return "Hello From C#!";
        }

        public UserData login(string username, string password)
        {
            return ConvertToUserData(_modelUser.GetAll().Find(user =>
                user.Name.Equals(username) && user.Password.Equals(password)));
        }

        private bool Validate(string username, string password, string confirm)
        {
            return _modelUser.GetAll().Exists(user => !user.Name.Equals(username)) && password.Equals(confirm);
        }

        public UserData signUp(string username, string password, string confirm)
        {
            return ConvertToUserData(
                Validate(username, password, confirm)
                    ? _modelUser.FindElementById(_modelUser.Add(new User(username, password))) : null);
        }

        public List<PlayerData> getPlayers()
        {
            return _modelPlayer.GetAll().Select(ConvertToPlayerData).ToList();
        }

        public List<EventData> getEvents()
        {
            return _modelEvent.GetAll().Select(ConvertToEventData).ToList();
        }

        public List<EventData> getEventsFromPlayer(PlayerData player)
        {
            return _modelPlayer.GetAllPlayerEvents(ConvertToPlayer(player).GetId()).Select(ConvertToEventData).ToList();
        }

        public List<PlayerData> getPlayersFromEvent(EventData @event)
        {
            return _modelEvent.GetAllEventPlayers(ConvertToEvent(@event).GetId()).Select(ConvertToPlayerData).ToList();
        }

        public int addPlayer(PlayerData player, List<EventData> events)
        {
            var id = _modelPlayer.Add(ConvertToPlayer(player));
            events.ForEach(item => _modelEvent.AddEventToPlayer(item.Id, id));
            UpdateSubscribers(Notification.UPDATE_ALL);
            return id;
        }

        private void UpdateSubscribers(Notification notification)
        {
            _subscribers.ForEach(subscriber => subscriber.update(notification));
        }

        public void subscribe(string hostname, int port)
        {
            TTransport transport = new TSocket(hostname, port);
            transport.Open();
            TProtocol protocol = new TBinaryProtocol(transport);
            var client = new Subscriber.Client(protocol);
            _subscribers.Add(client);
            Console.WriteLine("Client#" + _subscribers.Count + " Connected!");
            UpdateSubscribers(Notification.UPDATE_ALL);
        }

        public void unsubscribe(string hostname, int port)
        {
            UpdateSubscribers(Notification.UPDATE_ALL);
        }
    }
}

File: Systems for Design and Implementation/event.manager.java/src/main/java/network/ConnectionManager.java
package network;

import domain.Event;
import domain.EventStyle;
import domain.Player;
import domain.User;
import org.apache.thrift.TException;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;
import service.*;

import java.util.*;
import java.util.stream.Collectors;

public class ConnectionManager extends Observable implements Connection {

    private final String host;
    private final Integer serverPort;
    private Integer clientPort;
    private Subsciption.Client connection;
    private User active;

    public ConnectionManager(String host, Integer port) throws TTransportException {
        this.host = host;
        this.serverPort = port;
        clientPort = new Random().nextInt(9999) + 1000;
        Runnable task = () -> {
            Listener listener = new Listener(clientPort);
            listener.addObserver(this);
            listener.run();
        };
        new Thread(task).start();
    }

    @Override
    public void start() throws TException {
        TTransport transport = new TSocket(host, serverPort);
        transport.open();
        TProtocol protocol = new TBinaryProtocol(transport);
        connection = new Subsciption.Client(protocol);
        connection.subscribe(host, clientPort);
    }

    @Override
    public void setUser(User user) {
        this.active = user;
    }

    @Override
    public User getUser() {
        return active;
    }

    @Override
    public void popObserver(Observer observer) {
        removeObserver(observer);
    }

    @Override
    public void pushObserver(Observer observer) {
        addObserver(observer);
    }

    @Override
    public void update(Notification notification) {
        notifyObservers(notification);
    }

    private UserData convertToUserData(User user) {
        return new UserData(user.getName(), user.getPassword());
    }

    private User convertToUser(UserData user) {
        return new User(user.id, user.username, user.password);
    }

    private PlayerData convertToPlayerData(Player player) {
        return new PlayerData(player.getName(), player.getAge());
    }

    private Player convertToPlayer(PlayerData player) {
        return new Player(player.id, player.name, player.age);
    }

    private EventData convertToEventData(Event event) {
        return new EventData(event.getDistance(), event.getStyle().toString());
    }

    private Event convertToEvent(EventData event) {
        return new Event(event.id, event.distance, EventStyle.fromString(event.style));
    }

    @Override
    public User login(String username, String password) throws TException {
        return convertToUser(connection.login(username, password));
    }

    @Override
    public User signUp(String username, String password, String confirm) throws TException {
        return convertToUser(connection.signUp(username, password, confirm));
    }

    @Override
    public List<Player> getPlayers() throws TException {
        return connection.getPlayers().stream().map(this::convertToPlayer).collect(Collectors.toList());
    }

    @Override
    public List<Event> getEventsFromPlayer(Player player) throws TException {
        return connection.getEventsFromPlayer(convertToPlayerData(player))
                .stream().map(this::convertToEvent).collect(Collectors.toList());
}

    @Override
    public List<Event> getEvents() throws TException {
        return connection.getEvents().stream().map(this::convertToEvent).collect(Collectors.toList());
    }

    @Override
    public List<Player> getPlayerFromEvent(Event event) throws TException {
        return connection.getPlayersFromEvent(convertToEventData(event))
                .stream().map(this::convertToPlayer).collect(Collectors.toList());
    }

    @Override
    public Integer addPlayer(String name, Integer age, List<Event> events) throws TException {
        return connection.addPlayer(convertToPlayerData(new Player(name, age)),
                events.stream().map(this::convertToEventData).collect(Collectors.toList()));
    }

    @Override
    public void stop() throws TException {
        connection.unsubscribe(host, clientPort);
    }
}
